Index: /src/examples/1_tokio_setup/src/main.rs
===================================================================
--- /src/examples/1_tokio_setup/src/main.rs
+++ /src/examples/1_tokio_setup/src/main.rs
@@ -1,3 +1,16 @@
-fn main() {
-    println!("Hello, world!");
+async fn say_world() {
+    println!("world");
 }
+
+// We need to add the #[tokio::main] attribute to the main function to use Tokio.
+#[tokio::main]
+async fn main() {
+    // Calling `say_world()` does not execute the body of `say_world()`.
+    let op = say_world();
+
+    // This println! comes first
+    println!("hello");
+
+    // Calling `.await` on `op` starts executing `say_world`.
+    op.await;
+}
